### PR TITLE
CircleCI to Github actions migration for Kedro-Viz (test)

### PR DESCRIPTION
**Pull request by jitu5**
## Description

Resolves [#1785](https://github.com/kedro-org/kedro-viz/issues/1785)

### Reusable Workflows:
*  #### `lint.yml`:
    * Triggered by another workflow call : `all-checks.yml` & `build-backend.yml`
    * Inputs: os & python-version
    * Checks out code, installs dependencies and runs linter.

*  #### `e2e-tests.yml`:
    * Triggered by another workflow call : `all-checks.yml` & `build-backend.yml`
    * Inputs: os, python-version, branch
    * Checks out code, installs dependencies and runs end to end tests.

*  #### `unit-tests.yml`:
    * Triggered by another workflow call : `all-checks.yml` & `build-backend.yml`
    * Inputs: os, python-version, branch
    * Checks out code, installs dependencies and runs unit-tests.

*  #### `javascript-lint-and-tests.yml`:
    * Triggered by:
        * Call by another workflow : `all-checks.yml`
        * PRs on any branch, excluding those with changes to files in the `package/**` paths.
        * Pushes on any branch, excluding those with changes to files in the `package/**` paths.
    * Checks out code, installs dependencies and javascript linter & tests.

*  #### `all-checks.yml`:
    * Triggered by calls from another workflow : `check-release.yml` (will be added in separate PR)
    * Run every day at 1:00 AM(UTC time)
    * Calls - `unit-test.yml`, `lint.yml`, `e2e-tests.yml` and `javascript-lint-and-tests.yml` as separate jobs.

### Reusable Actions:

*  #### `install_node_dependencies/action.yml`:
    * Triggered by calls from workflow : `e2e-tests.yml`,  `unit-tests.yml`, `javascript-lint-and-tests.yml` and `release.yml` (will be added in separate PR)
    * Sets up a specific Node.js version, caches Node modules, and installs Node dependencies.

*  #### `install_python_dependencies/action.yml`:
    * Triggered by calls from workflow :  `e2e-tests.yml`,  `unit-tests.yml`, `javascript-lint-and-tests.yml` and `lint.yml`
    * Installs Kedro from the main branch and other Python dependencies, then prints the Python version and installed packages.

*  #### `setup_tests/action.yml`:
    * Triggered by calls from workflow :  `e2e-tests.yml` and `unit-tests.yml`.
    * Sets up the testing environment by setting up Python and Node.js, caching Python packages, installing Kedro and other Python dependencies, and building the React application.

### Workflows:

*  #### `build-backend.yml`:
    * Triggered by:
        * PRs on any branch with files having `package/**` & `.github/**` paths
        * Push on any branch with files having `package/**` & `.github/**` paths
    * Calls - `unit-test.yml`, `lint.yml` and `e2e-tests.yml` as separate jobs.

Screenshot 2024-04-11 at 1 56 54 p m

[Miro board link](https://miro.com/app/board/uXjVKZUQz8Y=/)

#### The following will be added in separate PRs

* Check Release (Workflow)
* Release (Workflow)
* Deploy Demo (Workflow)

#### Notes 
- The lint test is only run on Python 3.9 - This is what we had on CircleCI.
- I've added `cypress/fixtures/graphql/` to `trufflehog-ignore.txt` because secret-scan was complaining some files under it.
- `unit-tests.yml` & `e2e-tests.yml` run on main and demo branches for Windows and all branch for ubuntu, This is what we had on CircleCI.

## Checklist

- [X] Read the [contributing](/CONTRIBUTING.md) guidelines
- [X] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added new entries to the `RELEASE.md` file
- [ ] Added tests to cover my changes

### CODE REVIEW

File: .github/workflows/all-checks.yml
Comment (SajidAlamQB) [inline .github/workflows/all-checks.yml:36]: We can give this directly to `os:` if its the only OS needed, we don't need the matrix if that is the case.

File: .github/workflows/all-checks.yml
Comment (SajidAlamQB) [inline .github/workflows/all-checks.yml:51]: Same here, the matrix is not needed if they are only single inputs we can feed them directly.

File: .github/actions/install_node_dependencies/action.yml
Comment (SajidAlamQB) [inline .github/actions/install_node_dependencies/action.yml:5]: Haven't used composite before this looks great. 👍

**Review by SajidAlamQB — COMMENTED** [review]
Hey Jitendra, just had a look over the draft it looks great I've just left some minor comments. Happy to approve wants it out of draft otherwise.

File: .github/workflows/e2e-tests.yml
Comment (ankatiyar) [inline .github/workflows/e2e-tests.yml:16]: Just out of curiosity, is there a reason it was set up like this on CircleCI?

File: .github/workflows/all-checks.yml
Comment (ankatiyar) [reply .github/workflows/all-checks.yml:36]: Agree with this!

Comment (ankatiyar) [reply .github/workflows/all-checks.yml:36]: Agree with this!

File: .github/workflows/all-checks.yml
Comment (jitu5) [reply .github/workflows/all-checks.yml]: Updated!

File: .github/workflows/e2e-tests.yml
Comment (jitu5) [reply .github/workflows/e2e-tests.yml:16]: @ankatiyar  Because testing on Windows is time-consuming, we have decided for simplicity to only test on the main and demo versions. Therefore, for each commit, we do not perform full testing; instead, we conduct it when merging to the main and demo branches.

Comment (ankatiyar) [reply .github/workflows/e2e-tests.yml:16]: Makes sense!

**Review by ankatiyar — COMMENTED** [review]
I think you can combine `build-frontend.yml` and `javascript-lint-and-tests.yml` into one if `build-frontend.yml` is just calling `javascript-lint-and-tests.yml`. Otherwise everything looks good!

**Review by ankatiyar — APPROVED** [review]
LGTM! Great work @jitu5 💯

File: .github/actions/install_node_dependencies/action.yml
Comment (ravi-kumar-pilla) [inline .github/actions/install_node_dependencies/action.yml:9]: [Nit] To make this reusable may be we can have node-version as input ?

File: .github/actions/install_node_dependencies/action.yml
Comment (ravi-kumar-pilla) [inline .github/actions/install_node_dependencies/action.yml:21]: [Nit] To make this reusable may be we can have path to package.json as input which defaults to current directory

File: .github/actions/install_python_dependencies/action.yml
Comment (ravi-kumar-pilla) [inline .github/actions/install_python_dependencies/action.yml:1]: [Nit] For our project it make sense to have kedro installed here. But in a general case as the file name says - we should only install python dependencies and have kedro installation in a separate step. But I am okay if the team feels otherwise.

File: .github/workflows/e2e-tests.yml
Comment (ravi-kumar-pilla) [inline .github/workflows/e2e-tests.yml:51]: -name: **Run all end to end tests** ?

File: .github/workflows/unit-tests.yml
Comment (ravi-kumar-pilla) [inline .github/workflows/unit-tests.yml:49]: I know this ticket is only about migrating CircleCI to github actions. I would like to know if you came across a way to reuse some setup here. If we see `make e2e-tests` and `make pytest`, except the last step entire job is the same. If we can some way re-use the steps it would be nice. @jitu5

File: .github/workflows/e2e-tests.yml
Comment (jitu5) [reply .github/workflows/e2e-tests.yml:16]: @ravi-kumar-pilla this condition ensures that the tests will run on Ubuntu regardless of the branch, and for Windows, the tests will only run if the branch is either "main" or "demo".  So for PRs to main this will works regardless of the OS.

File: .github/actions/install_python_dependencies/action.yml
Comment (jitu5) [reply .github/actions/install_python_dependencies/action.yml:1]: @ravi-kumar-pilla 
If you notice workflows where this action is used in its step name, it says ‘Install Kedro and other Python Dependencies.’ I want to avoid repeating the step of ‘Install Kedro’. But I can rename the action file to ‘install_kedro_and_python_dependencies’ for more clarity.

File: .github/workflows/unit-tests.yml
Comment (jitu5) [reply .github/workflows/unit-tests.yml]: @ravi-kumar-pilla Yes, I moved the common steps to reusable "Setup Tests" action, which can be reuse in both  e2e-test and unit-tests workflows.

File: .github/workflows/e2e-tests.yml
Comment (jitu5) [reply .github/workflows/e2e-tests.yml]: Done.

File: .github/actions/install_node_dependencies/action.yml
Comment (jitu5) [reply .github/actions/install_node_dependencies/action.yml]: Done.

File: .github/workflows/e2e-tests.yml
Comment (ravi-kumar-pilla) [reply .github/workflows/e2e-tests.yml:16]: Yes, #1853 has issue on windows. So the tests failed after merging to main. So, I was referring to running the tests on windows (atleast 1 python version) to be safe when a PR is raised to main

File: .github/actions/install_python_dependencies/action.yml
Comment (ravi-kumar-pilla) [reply .github/actions/install_python_dependencies/action.yml:1]: `install_kedro_and_python_dependencies` this is more clear but also a bit lengthy. I would suggest to have `install_kedro_and_python_dependencies` for better readability

File: .github/workflows/e2e-tests.yml
Comment (jitu5) [reply .github/workflows/e2e-tests.yml:16]: @ravi-kumar-pilla Okay got it.
This workflow(e2e-tests.yml) and parent workflow (build-backend.yml) will be trigger when a PR is raised to main for windows as well.

File: .github/actions/install_node_dependencies/action.yml
Comment (ravi-kumar-pilla) [inline .github/actions/install_node_dependencies/action.yml:7]: I think we have a default value, so this should be optional right ?

File: .github/actions/setup_tests/action.yml
Comment (ravi-kumar-pilla) [inline .github/actions/setup_tests/action.yml:7]: this should be optional considering we have a default value

File: .github/actions/setup_tests/action.yml
Comment (ravi-kumar-pilla) [inline .github/actions/setup_tests/action.yml:11]: this should be optional considering we have a default value

File: .github/actions/setup_tests/action.yml
Comment (ravi-kumar-pilla) [inline .github/actions/setup_tests/action.yml:1]: This is awesome !!

File: .github/actions/install_node_dependencies/action.yml
Comment (jitu5) [reply .github/actions/install_node_dependencies/action.yml]: Yes, updated.

File: .github/actions/setup_tests/action.yml
Comment (jitu5) [reply .github/actions/setup_tests/action.yml]: Yes, updated.

**Review by SajidAlamQB — APPROVED** [review]
Thank you @jitu5, this looks great!

File: .github/actions/install_node_dependencies/action.yml
Comment (rashidakanchwala) [inline .github/actions/install_node_dependencies/action.yml:8]: we should update this to 18

Comment (jitu5) [reply .github/actions/install_node_dependencies/action.yml]: Done.

File: .github/workflows/e2e-tests.yml
Comment (rashidakanchwala) [reply .github/workflows/e2e-tests.yml]: @jitu5 , @ravi-kumar-pilla  I suggest we stick to our previous approach. Whether we run one Windows test or multiple tests in parallel, they take approximately the same amount of time, around 30 minutes. This duration is double that of running Python tests. Waiting for 30 minutes for tests to complete after pushing to a PR wasn't a great dev experience, hence this change was introduced. Windows errors are usually rare, so it's ok to address them reactively once the changes are merged into the main branch.

File: .github/workflows/merge-gatekeeper.yml
Comment (ravi-kumar-pilla) [inline .github/workflows/merge-gatekeeper.yml:1]: this file seems new, can you provide a bit of context on why this is needed. Thank you @jitu5

Comment (jitu5) [reply .github/workflows/merge-gatekeeper.yml:1]: @ravi-kumar-pilla 
The merge-gatekeeper workflow is triggered by pull requests on the main branch. It checks every 30 seconds to see if other specified workflows have completed and passed. If any of these workflows fail, the merge-gatekeeper will also fail. This workflow must pass for pull requests to be mergeable. Also, it is configured to ignore certain CircleCI workflows as specified in the ignored field. 
once this PR is merge will need to mark this workflow as required.
Screenshot 2024-04-18 at 10 18 49 a m

**Review by ravi-kumar-pilla — APPROVED** [review]
Awesome work @jitu5 !! Excited to have gh-actions 🥇 💯

**Review by rashidakanchwala — APPROVED** [review]
Amazing LGTM!! feel free to merge once tests pass. thanks @jitu5

**Pull request merged**
